File: src/KubeOps.Transpiler/Rbac.fs
module KubeOps.Transpiler.Rbac

open System
open System.Linq
open KubeOps.Abstractions.Rbac
open k8s.Models

let private convertToStrings verbs =
    let values =
        match verbs with
        | RbacVerb.None -> Seq.empty
        | _ when verbs.HasFlag(RbacVerb.All) -> seq { "*" }
        | _ ->
#if NETSTANDARD
            Enum.GetValues(typeof<RbacVerb>).Cast<RbacVerb>()
#else
            Enum.GetValues<RbacVerb>()
#endif
            |> Seq.filter (fun x -> x <> RbacVerb.None && x <> RbacVerb.All)
            |> Seq.filter verbs.HasFlag
            |> Seq.map (fun x -> x.ToString().ToLower())

    values.ToList()

let private transpileGenericAttribute (a: GenericRbacAttribute) =
    V1PolicyRule(
        ApiGroups = a.Groups,
        Resources = a.Resources,
        NonResourceURLs = a.Urls,
        Verbs = convertToStrings a.Verbs
    )

let private transpileEntityAttributes (attributes: EntityRbacAttribute seq) =
    attributes
    |> Seq.collect (fun a -> a.Entities |> Seq.map (fun e -> (e, a.Verbs)))
    |> Seq.groupBy fst
    |> Seq.map (fun (e, g) ->
        (Entities.ToEntityMetadata e, g |> Seq.map snd |> Seq.fold (fun acc verb -> acc ||| verb) RbacVerb.None))
    |> Seq.groupBy snd
    |> Seq.map (fun (verbs, group) -> (verbs, group |> Seq.map fst |> Seq.map fst))
    |> Seq.map (fun (verbs, crds) ->
        V1PolicyRule(
            ApiGroups = (crds |> Seq.map (fun x -> x.Group) |> Seq.distinct).ToList(),
            Resources = (crds |> Seq.map (fun x -> x.Plural) |> Seq.distinct).ToList(),
            Verbs = convertToStrings verbs
        ))

let private transpileEntityStatusAttributes (attributes: EntityRbacAttribute seq) =
    attributes
    |> Seq.collect (fun a -> a.Entities |> Seq.map (fun e -> (e, a.Verbs)))
    |> Seq.filter (fun (e, _) -> e.GetProperty("Status") <> null)
    |> Seq.groupBy fst
    |> Seq.map fst
    |> Seq.map Entities.ToEntityMetadata
    |> Seq.map fst
    |> Seq.map (fun e ->
        V1PolicyRule(
            ApiGroups = [ e.Group ].ToList(),
            Resources = [ $"{e.Plural}/status" ].ToList(),
            Verbs = convertToStrings (RbacVerb.Get ||| RbacVerb.Patch ||| RbacVerb.Update)
        ))

let Transpile (attributes: RbacAttribute seq) =
    let generic =
        attributes
        |> Seq.filter (fun x -> x :? GenericRbacAttribute)
        |> Seq.cast<GenericRbacAttribute>
        |> Seq.map transpileGenericAttribute

    let entityAttrs =
        attributes
        |> Seq.filter (fun x -> x :? EntityRbacAttribute)
        |> Seq.cast<EntityRbacAttribute>

    let entity = transpileEntityAttributes entityAttrs
    let entityStatus = transpileEntityStatusAttributes entityAttrs

    Seq.concat [ generic; entity; entityStatus ]

File: src/KubeOps.Transpiler/Rbac.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the Apache 2.0 License.
// See the LICENSE file in the project root for more information.

using System.Reflection;

using k8s.Models;

using KubeOps.Abstractions.Rbac;

namespace KubeOps.Transpiler;

/// <summary>
/// Transpiler for Kubernetes RBAC attributes to create <see cref="V1PolicyRule"/>s.
/// </summary>
public static class Rbac
{
    /// <summary>
    /// Convert a list of <see cref="RbacAttribute"/>s to a list of <see cref="V1PolicyRule"/>s.
    /// The rules are grouped by entity type and verbs.
    /// </summary>
    /// <param name="context">The <see cref="MetadataLoadContext"/> that was used to load the attributes.</param>
    /// <param name="attributes">List of <see cref="RbacAttribute"/>s.</param>
    /// <returns>A converted, grouped list of <see cref="V1PolicyRule"/>s.</returns>
    public static IEnumerable<V1PolicyRule> Transpile(
        this MetadataLoadContext context,
        IEnumerable<CustomAttributeData> attributes)
    {
        var list = attributes.ToList();

        var generic = list
            .Where(a => a.AttributeType == context.GetContextType<GenericRbacAttribute>())
            .Select(a => new V1PolicyRule
            {
                ApiGroups = a.GetCustomAttributeNamedArrayArg<string>(nameof(GenericRbacAttribute.Groups)),
                Resources = a.GetCustomAttributeNamedArrayArg<string>(nameof(GenericRbacAttribute.Resources)),
                NonResourceURLs = a.GetCustomAttributeNamedArrayArg<string>(nameof(GenericRbacAttribute.Urls)),
                Verbs = ConvertToStrings(
                    a.GetCustomAttributeNamedArg<RbacVerb>(context, nameof(GenericRbacAttribute.Verbs))),
            });

        var entities = list
            .Where(a => a.AttributeType == context.GetContextType<EntityRbacAttribute>())
            .SelectMany(attribute =>
                attribute.GetCustomAttributeCtorArrayArg<Type>(0).Select(type =>
                    (EntityType: type,
                        Verbs: attribute.GetCustomAttributeNamedArg<RbacVerb>(
                            context,
                            nameof(GenericRbacAttribute.Verbs)))))
            .GroupBy(e => e.EntityType)
            .Select(
                group => (
                    Crd: context.ToEntityMetadata(group.Key),
                    Verbs: group.Aggregate(RbacVerb.None, (accumulator, element) => accumulator | element.Verbs)))
            .GroupBy(group => (group.Crd.Metadata.Group, group.Verbs))
            .Select(
                group => new V1PolicyRule
                {
                    ApiGroups = [group.Key.Group],
                    Resources = group.Select(crd => crd.Crd.Metadata.PluralName).Distinct().ToList(),
                    Verbs = ConvertToStrings(group.Key.Verbs),
                });

        var entityStatus = list
            .Where(a => a.AttributeType == context.GetContextType<EntityRbacAttribute>())
            .SelectMany(attribute =>
                attribute.GetCustomAttributeCtorArrayArg<Type>(0).Select(type =>
                    (EntityType: type,
                        Verbs: attribute.GetCustomAttributeNamedArg<RbacVerb>(
                            context,
                            nameof(GenericRbacAttribute.Verbs)))))
            .Where(e => e.EntityType.GetProperty("Status") != null)
            .GroupBy(e => e.EntityType)
            .Select(group => context.ToEntityMetadata(group.Key))
            .Select(
                crd => new V1PolicyRule
                {
                    ApiGroups = [crd.Metadata.Group],
                    Resources = [$"{crd.Metadata.PluralName}/status"],
                    Verbs = ConvertToStrings(RbacVerb.Get | RbacVerb.Patch | RbacVerb.Update),
                });

        return generic.Concat(entities).Concat(entityStatus);
    }

    private static string[] ConvertToStrings(RbacVerb verbs) => verbs switch
    {
        RbacVerb.None => Array.Empty<string>(),
        _ when verbs.HasFlag(RbacVerb.All) => ["*"],
        _ when verbs.HasFlag(RbacVerb.AllExplicit) =>
            Enum.GetValues<RbacVerb>()
                .Where(v => v != RbacVerb.All && v != RbacVerb.None && v != RbacVerb.AllExplicit)
                .Select(v => v.ToString().ToLowerInvariant())
                .ToArray(),
        _ =>
            Enum.GetValues<RbacVerb>()
                .Where(v => verbs.HasFlag(v) && v != RbacVerb.All && v != RbacVerb.None && v != RbacVerb.AllExplicit)
                .Select(v => v.ToString().ToLowerInvariant())
                .ToArray(),
    };
}
